File: face_recognition.py


import sys
import json
import face_recognition
import cv2
import numpy as np
from PIL import Image
import io
import base64

def encode_face(image_path):
    """Extract face encoding from image"""
    try:
        # Load image
        image = face_recognition.load_image_file(image_path)
        
        # Find face encodings
        face_encodings = face_recognition.face_encodings(image)
        
        if len(face_encodings) > 0:
            # Return the first face encoding found
            encoding = face_encodings[0].tolist()
            return json.dumps({
                'success': True,
                'encoding': encoding
            })
        else:
            return json.dumps({
                'success': False,
                'error': 'No face found in image'
            })
    except Exception as e:
        return json.dumps({
            'success': False,
            'error': str(e)
        })

def compare_faces(data_file):
    """Compare two face encodings"""
    try:
        with open(data_file, 'r') as f:
            data = json.load(f)
        
        encoding1 = np.array(data['encoding1'])
        encoding2 = np.array(data['encoding2'])
        tolerance = data.get('tolerance', 0.6)
        
        # Calculate face distance
        distance = face_recognition.face_distance([encoding1], encoding2)[0]
        
        # Convert distance to confidence
        confidence = max(0, 1 - distance)
        is_match = distance <= tolerance
        
        return json.dumps({
            'success': True,
            'is_match': is_match,
            'confidence': confidence,
            'distance': distance
        })
    except Exception as e:
        return json.dumps({
            'success': False,
            'error': str(e)
        })

if __name__ == "__main__":
    if len(sys.argv) < 2:
        print(json.dumps({'error': 'No command specified'}))
        sys.exit(1)
    
    command = sys.argv[1]
    
    if command == "encode" and len(sys.argv) >= 3:
        result = encode_face(sys.argv[2])
        print(result)
    elif command == "compare" and len(sys.argv) >= 3:
        result = compare_faces(sys.argv[2])
        print(result)
    else:
        print(json.dumps({'error': 'Invalid command or arguments'}))

File: face_recognition.php
<?php
class FaceRecognition {
    private $pythonScript;
    
    public function __construct() {
        $this->pythonScript = __DIR__ . '/face_recognition.py';
    }
    
    public function extractFaceEncoding($imagePath) {
        // Call Python script for face encoding (requires face_recognition library)
        $command = "python3 {$this->pythonScript} encode " . escapeshellarg($imagePath);
        $output = shell_exec($command);
        
        if($output) {
            return json_decode(trim($output), true);
        }
        
        return false;
    }
    
    public function compareFaces($encoding1, $encoding2, $tolerance = 0.6) {
        // Compare two face encodings
        $data = [
            'encoding1' => $encoding1,
            'encoding2' => $encoding2,
            'tolerance' => $tolerance
        ];
        
        $tempFile = tempnam(sys_get_temp_dir(), 'face_compare');
        file_put_contents($tempFile, json_encode($data));
        
        $command = "python3 {$this->pythonScript} compare " . escapeshellarg($tempFile);
        $output = shell_exec($command);
        
        unlink($tempFile);
        
        if($output) {
            return json_decode(trim($output), true);
        }
        
        return false;
    }
}